chart/chart.go

```diff
@@ -0,0 +1,214 @@
+	preDocs      string = "docs/"
+)
+
+// Chart represents a complete chart.
+//
+// A chart consists of the following parts:
+//
+// 	- Chart.yaml: In code, we refer to this as the Chartfile
+// 	- templates/*: The template directory
+// 	- README.md: Optional README file
+// 	- LICENSE: Optional license file
+// 	- hooks/: Optional hooks registry
+//	- docs/: Optional docs directory
+//
+// Packed charts are stored in gzipped tar archives (.tgz). Unpackaged charts
+// are directories where the directory name is the Chartfile.Name.
+//
+// Optionally, a chart might also locate a provenance (.prov) file that it
+// can use for cryptographic signing.
+type Chart interface {
+	// Chartfile resturns a *Chartfile for this chart.
+	Chartfile() *Chartfile
+	// Dir returns a directory where the chart can be accessed.
+	Dir() string
+
+	// Close cleans up a chart.
+	Close() error
+}
+
+type dirChart struct {
+	chartfile *Chartfile
+	dir       string
+}
+
+func (d *dirChart) Chartfile() *Chartfile {
+	return d.chartfile
+}
+
+func (d *dirChart) Dir() string {
+	return "."
+}
+
+func (d *dirChart) Close() error {
+	return nil
+}
+
+type tarChart struct {
+	chartfile *Chartfile
+	tmpDir    string
+}
+
+func (t *tarChart) Chartfile() *Chartfile {
+	return t.chartfile
+}
+
+func (t *tarChart) Dir() string {
+	return "."
+}
+
+func (t *tarChart) Close() error {
+	// Remove the temp directory.
+	return os.RemoveAll(t.tmpDir)
+}
+
+// LoadDir loads an entire chart from a directory.
+//
+// This includes the Chart.yaml (*Chartfile) and all of the manifests.
+//
+// If you are just reading the Chart.yaml file, it is substantially more
+// performant to use LoadChartfile.
+func LoadDir(chart string) (Chart, error) {
+	if fi, err := os.Stat(chart); err != nil {
+		return nil, err
+	} else if !fi.IsDir() {
+		return nil, fmt.Errorf("Chart %s is not a directory.", chart)
+	}
+
+	cf, err := LoadChartfile(filepath.Join(chart, "Chart.yaml"))
+	if err != nil {
+		return nil, err
+	}
+
+	c := &dirChart{
+		chartfile: cf,
+		dir:       chart,
+	}
+
+	return c, nil
+}
+
+// Load loads a chart from a chart archive.
+//
+// A chart archive is a gzipped tar archive that follows the Chart format
+// specification.
+func Load(archive string) (Chart, error) {
+	if fi, err := os.Stat(archive); err != nil {
+		return nil, err
+	} else if fi.IsDir() {
+		return nil, errors.New("cannot load a directory with chart.Load()")
+	}
+
+	raw, err := os.Open(archive)
+	if err != nil {
+		return nil, err
+	}
+	defer raw.Close()
+
+	unzipped, err := gzip.NewReader(raw)
+	if err != nil {
+		return nil, err
+	}
+	defer unzipped.Close()
+
+	untarred := tar.NewReader(unzipped)
+	c, err := loadTar(untarred)
+	if err != nil {
+		return c, err
+	}
+
+	cf, err := LoadChartfile(filepath.Join(c.tmpDir, ChartfileName))
+	if err != nil {
+		return c, err
+	}
+	c.chartfile = cf
+	return c, nil
+}
+
+func loadTar(r *tar.Reader) (*tarChart, error) {
+	td, err := ioutil.TempDir("", "chart-")
```

Comment:
The package here implements two backends for a chart: One is the uncompressed, unpackaged directory. The other is the tar (which does use `archive/tar`). Both adhere to the same interface.

For a tar file, the code dumps the data into a temp directory. There are two major reasons to do this the way it's done here:

1. Given the new format, the sizes of charts are less predictable, so purely from the memory management angle, we didn't want to keep everything in memory. (To wit, Sebastian just suggested adding SVG image files to charts.) Nor does it make a lot of sense to keep a bunch of readers open, and try to manage concurrency with those.
2. Given what we discussed about generation headers, we now perform a number of operations on directories and on files (as path-relative) in directories. We also now explicitly allow generators to create files inside of the package. All of these are more easily accomplished by extracting the archive to a temp directory and then allowing them to be modified as files on a filesystem.

Maybe I didn't make it clear what I'm trying to do with this package. This is supposed to be the standard package for working with charts. Special purpose chart handling done elsewhere (like the expandybird stuff) should be refactored to use this. The client code will be using this as well. So I'm not advocating having duplicate code. I'm advocating rewriting all of the other code to use this one implementation, which adheres to the standard we just all agreed upon.

FWIW, a follow-up PR that I am working on has the code necessary for saving a chart, and also packing a chart into a tarball. I just didn't want to add a huge PR with everything jammed in it.